File: Python/89.Gray Code.py
#求n位数的格雷码(二进制中每次更改一位)
class Solution(object):
    def grayCode(self, n):
        tmp , result = 0 , [0]
        for x in range(2**n-1):
            for i in range(n):
                t = tmp^(2**i)
                if not t in result:
                    result.append(t)
                    tmp = t
                    break
        return result

s = Solution()
print(s.grayCode(3))

File: C++/Backtracking/89.Gray Code.cpp
/*
格雷码，给出位数n，按顺序返回格雷码。
顺序为，二进制位每次变更一位。
*/
/*
思路：第k个格雷码的求法位k ^(k/2)。
*/
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> grayCode(int n) {
    	vector<int>	result;
        for(int i=0; i < 1<<n; ++i)
        {
        	result.push_back(i ^ i>>1);
        }
        return result;
    }
};
int main()
{
	Solution s;
    vector<int> out = s.grayCode(3);
    for(int n:out)
    	cout<<n<<endl;
}

